File: 23_设计类/1348. 推文计数.ts
type TweetName = string
type Time = number
type Freq = 'minute' | 'hour' | 'day'

class TweetCounts {
  private tweetMap: Map<TweetName, Time[]>
  constructor() {
    this.tweetMap = new Map()
  }

  /**
   *
   * @param tweetName
   * @param time
   * 用户 tweetName 在 time（以 秒 为单位）时刻发布了一条推文。
   */
  recordTweet(tweetName: string, time: number): void {
    !this.tweetMap.has(tweetName) && this.tweetMap.set(tweetName, [])
    this.tweetMap.get(tweetName)!.push(time)
  }

  /**
   *
   * @param freq
   * @param tweetName
   * @param startTime
   * @param endTime
   * 返回从开始时间 startTime（以 秒 为单位）到结束时间 endTime（以 秒 为单位）内，
   * 每 分 minute，时 hour 或者 日 day （取决于 freq）内指定用户 tweetName 发布的推文总数。
   */
  getTweetCountsPerFrequency(
    freq: Freq,
    tweetName: string,
    startTime: number,
    endTime: number
  ): number[] {
    endTime++
    let length: number
    if (freq === 'minute') length = 60
    else if (freq === 'hour') length = 60 * 60
    else length = 24 * 60 * 60

    // 有点问题
    const res = Array(~~((endTime - startTime - 1) / (length + 1))).fill(0)
    for (const time of this.tweetMap.get(tweetName) || []) {
      if (time >= startTime && time < endTime) {
        res[~~((time - startTime) / length)]++
      }
    }

    return res
  }
}

export {}

/**
 * Your TweetCounts object will be instantiated and called as such:
 * var obj = new TweetCounts()
 * obj.recordTweet(tweetName,time)
 * var param_2 = obj.getTweetCountsPerFrequency(freq,tweetName,startTime,endTime)
 */

//  可以将每个用户的推文时间存储方式换成更有效的平衡二叉树。
//  与暴力法所使用的线性表相比，
//  平衡二叉树保证其中的元素使用二叉树有序排列，其时间复杂度与线性表的区别为：
// 对于插入操作，线性表的时间复杂度为 O(1)，平衡二叉树的时间复杂度为 O(logn)。
// 对于查询操作，线性表的时间复杂度为 O(n)，平衡二叉树的时间复杂度为 O(logn)。
// 即 Map<TweetName,TreeSet<Time>> 其中treeset有bisectLeft/Right查找第几个的api

File: 5_map/邻接表/1348. 推文计数.py
from typing import List
from collections import defaultdict
from sortedcontainers import SortedList


class TweetCounts:
    """一家社交媒体公司正试图通过分析特定时间段内出现的推文数量来监控其网站上的活动。"""

    def __init__(self):
        self.tweets = defaultdict(SortedList)

    def recordTweet(self, tweetName: str, time: int) -> None:
        """记录推文发布情况：用户 tweetName 在 time（以 秒 为单位）时刻发布了一条推文。"""
        self.tweets[tweetName].add(time)

    def getTweetCountsPerFrequency(
        self, freq: str, tweetName: str, startTime: int, endTime: int
    ) -> List[int]:
        """每个 时间块 中带有 tweetName 的 tweet 的数量"""
        if freq == "minute":
            seconds = 60
        elif freq == "hour":
            seconds = 60 * 60
        else:
            seconds = 60 * 60 * 24

        res = [0] * ((endTime - startTime) // seconds + 1)
        sl = self.tweets[tweetName]
        left, right = sl.bisect_left(startTime), sl.bisect_right(endTime) - 1
        for pos in range(left, right + 1):
            res[(sl[pos] - startTime) // seconds] += 1
        return res


if __name__ == '__main__':
    tweetCounts = TweetCounts()
    tweetCounts.recordTweet("tweet3", 0)
    tweetCounts.recordTweet("tweet3", 60)
    tweetCounts.recordTweet("tweet3", 10)
    tweetCounts.recordTweet("tweet3", 120)
    tweetCounts.recordTweet("tweet3", 160)
    print(tweetCounts.getTweetCountsPerFrequency("minute", "tweet3", 0, 210))
